src/feature/hs/hs_ob.c

```diff
@@ -218,3 +239,54 @@ hs_ob_parse_config_file(hs_service_config_t *config, const char *fname)
   tor_free(content);
   return ret;
 }
+
+/** Compute all possible subcredentials for every onion master key in the
+ * given service config object. The subcredentials is allocated and set as an
+ * continous array containing all possible values.
+ *
+ * The number of subcredential set in the array is returned (length). If the
+ * configuration has no OB master keys configured, 0 is returned and
+ * subcredentials is set to NULL. */
+size_t
+hs_ob_get_subcredentials(const hs_service_config_t *config,
+                         uint8_t **subcredentials)
+{
+  unsigned int num_pkeys, idx = 0;
+  int steps[3] = {0, -1, 1};
+  const uint64_t tp = hs_get_time_period_num(0);
+
+  tor_assert(config);
+
+  num_pkeys = smartlist_len(config->ob_master_pubkeys);
+  if (!num_pkeys) {
+    *subcredentials = NULL;
+    goto end;
+  }
+
+  /* Next loop will build 3 subcredentials for each public key. */
+  *subcredentials = tor_malloc_zero(DIGEST256_LEN * 3 * num_pkeys);
+
+  /* Time to build all the subcredentials for each time period: the previous
+   * one (-1), the current one (0) and the next one (1) in order to accomodate
+   * client and service consensus skew.
+   *
+   * If the client consensus after_time is at 23:00 but the service one is at
+   * 01:00, the client will be using the previous time period where the
+   * service will think it is the client next time period. Thus why we have
+   * to try them all.
+   *
+   * As a slight optimization, statistically, the current time period (0) will
+   * be the one to work first so we'll put them first in the array to maximize
+   * our chance of success. */
+
+  for (unsigned int i = 0; i <= sizeof(steps); i++) {
+    SMARTLIST_FOREACH_BEGIN(config->ob_master_pubkeys,
+                            const ed25519_public_key_t *, pkey) {
+      build_subcredential(pkey, tp + steps[i], subcredentials[idx]);
```

Comment:
I'm pretty sure this is going to crash.  You're treating subcredentials as an array of subcredentials, but you allocated an array of uint8_t above.